Update de.flapdoodle.embed.mongo to 4.13.0 in main (#297)
input-output-hk/atala-prism-mediator: README.md

| Live Demo                        | Discord                          | Atala PRISM Website            |
|----------------------------------|----------------------------------|--------------------------------|
| [![Atala-Symbol]][Link-LiveDemo] | [![Badge-Discord]][Link-Discord] | [![Atala-Favicon]][Link-Atala] |

| CI              | Project Stage  | Commits since latest release       |
|-----------------|----------------|------------------------------------|
| ![CI][Badge-CI] | ![Badge-Stage] | ![Badge-CommitsSinceLatestRelease] |

# Mediator

A DID Comm v2 mediator  
A service that receives messages for many agents at a single endpoint and stores them with privacy.
A cloud-based agent that forwards messages to mobile devices.

```mermaid
graph LR
  A((Sender)) -- forward --> M((Mediator))
  M--pickup-->D((Reciever))
```

 - **CI** automates builds and tests that push to the main branch for all PRs created.
 - **Scala Steward** automates the creation of pull requests for libraries with updated dependencies, saving maintainers time and effort. It can also help keep libraries updated, improving their reliability and performance.

---

**#identus-mediator on Discord:**

> For the fastest answers, join the [#identus-mediator][Link-Discord] channel in the official Atala Discord and ask your questions, or chat with other Atala developers and pioneers!

Reviewer: This should link to the HL Discord

- [LICENSE](LICENSE) - Apache License, Version 2.0
- [Mediation Flows](Mediation-Flows.md) - Examples of mediation flows
- [Mediator Purpose](#Description)
- [Protocols Supported](#Protocols)
  - [Mediator Protocol State and Flow](Coordinate-Mediation-Protocol.md)
- [Pre-reqs](#pre-reqs)
- [Getting started](#getting-started)
  - [Docker only](#docker-only)
- [Configuring the mediator](#configure-the-mediator)
    - [Mediator identity](#identity)
    - [Mediator storage](#mediator-storage)
      - [Mediator-Storage-In-cloud](#mongodb-in-cloud) 
- [Mediator Deployment](#deploy)
- [Mediator Test suite](#mediator-tests)
- [Mediator Error Handling](Mediator-Error_Handling.md#error-handling)  
- [Mediator protocols state flow with problem reporting](Mediator-Error_Handling.md#problem-reports-in-mediator-flow)

## Description

DID Comm v2 (Decentralized Identifiers Communication Version 2) is a protocol engineered for secure, private, and decentralized communications between various entities utilizing decentralized identifiers (DIDs). A DID Comm v2 Mediator functions as an intermediary in the communication process, streamlining the exchange of messages among the involved parties.

> - Establishing Logical Connections - The Mediator empowers entities, which could be individuals or organizations, to forge secure connections amongst themselves. Each entity possesses a unique DID that acts as its identifier within the decentralized network.

Reviewer: ```suggestion
- **Establishing Logical Connections**: The Mediator empowers entities, which could be individuals or organizations, to forge secure connections amongst themselves. Each entity possesses a unique DID that acts as its identifier within the decentralized network.
```

> - DID resolution - When an entity seeks to communicate with another, it resolves the recipient's DID to procure the information necessary to establish a connection. This resolution procedure entails retrieving the recipient's public key and correlated metadata from a decentralized identity infrastructure, which could be a blockchain or distributed ledger.

Reviewer: ```suggestion
- **DID resolution**: When an entity seeks to communicate with another, it resolves the recipient's DID to procure the information necessary to establish a connection. This resolution entails retrieving the recipient's public key and correlated metadata from a decentralized identity infrastructure, a blockchain, or a distributed ledger.
```

> - Message encryption - The sender employs a double encryption technique for the message: initially for the ultimate receiver, and subsequently encapsulates the encrypted message within another encryption layer for the Mediator. This is achieved using the public keys of both the Mediator and the recipient obtained through the DID resolution process. Dual encryption ensures that only the intended recipient has the capacity to decrypt and access the message.

Reviewer: ```suggestion
- **Message encryption**: The sender employs a double encryption technique for the message, initially for the ultimate receiver, and subsequently encapsulates the encrypted message within another encryption layer for the Mediator. This procedure is possible using the public keys of both the Mediator and the recipient obtained through the DID resolution process. Dual encryption ensures that only the intended recipient can decrypt and access the message.
```

```mermaid
graph LR
  subgraph Encrypted message to Mediator
    subgraph Encrypted message to Reciever
    id1[[The plaintext message]]
    end
  end
```

> - Message routing - The sender transmits an encrypted message to the Mediator, which serves as a routing agent. In this role, the Mediator receives messages from the sender, decrypts one layer, and forwards them to the appropriate recipient based on the recipient's DID.

Reviewer: ```suggestion
- **Message routing**: The sender transmits an encrypted message to the Mediator, a routing agent. In this role, the Mediator receives messages from the sender, decrypts one layer, and forwards them to the appropriate recipient based on the recipient's DID.
```

> - Mediation process- The Mediator verifies the authenticity and integrity of the incoming message by checking the digital signature attached to it. This signature ensures that the message was indeed sent by the claimed sender and that it hasn't been tampered with during transmission.

Reviewer: ```suggestion
- **Mediation process**: The Mediator verifies the authenticity and integrity of the incoming message by checking the digital signature attached to it. This signature ensures that the claimed sender sent the message and hasn't been tampered with during transmission.
```

> - Message decryption - After verifying the message's authenticity, the Mediator decrypted one layer of the message using the mediator's private key, which is securely held by the mediator. Once decrypted, the next message becomes readable (the final plaintext intended for the final user it's still encrypted).

Reviewer: ```suggestion
- **Message decryption**: After verifying the message's authenticity, the mediator decrypted one layer of the message using the mediator's private key, which the mediator securely holds. Once decrypted, the following message becomes readable (the final plaintext intended for the final user is still encrypted).
```

> - Optional processing - The Mediator may perform additional processing on the message based on predefined rules or business logic. This could include applying filters, applying policies, or invoking external services.

Reviewer: ```suggestion
- **Optional processing**: The Mediator may perform additional processing on the message based on predefined rules or business logic, including applying filters, applying policies, or invoking external services.
```

> - Message forwarding - If necessary, the Mediator can further forward the decrypted message to additional entities in the communication flow. This enables multi-party communication scenarios.

Reviewer: ```suggestion
- **Message forwarding**: The mediator can forward the decrypted message to additional entities in the communication flow, enabling multi-party communication scenarios.
```

> By acting as an intermediary, the DID Comm v2 Mediator helps facilitate secure and private communication between entities while leveraging the decentralized nature of DIDs and cryptographic techniques to ensure the authenticity, integrity, and confidentiality of the messages exchanged.

Reviewer: ```suggestion
By acting as an intermediary, the DID Comm v2 Mediator helps facilitate secure and private communication between entities while leveraging the decentralized nature of DIDs and cryptographic techniques to ensure the message's authenticity, integrity, and confidentiality.
```

> The mediator is especially useful when the edge entities are not always online, like the mobile paradigm. Usually, we can assume that the mediator is always online.

Reviewer: ```suggestion
The mediator is particularly convenient for edge entities that are only sometimes online, such as in the mobile paradigm. It's a practical solution, as we can assume that the mediator is always online.
```
